### DISCUSSION TITLE
Segmentation fault for multi-threaded function calls and std::execution::par

### DISCUSSION THREAD

**Discussion by sashankmodali** [casadi/casadi#3351] [General]
Hello!

I am using CasADi to linearize an optimal control problem for a control horizon of 60 time steps. I would like to speed this up by using multithreaded function calls. But I'm not able to do so without causing segmentation faults. After reading [this comment](https://github.com/casadi/casadi/issues/2854#issuecomment-996562814), I have been trying to find a function definition that doesn't cause a segmentation fault. Could anyone help me with the correct function definition to use?

I've tried the following options in c++ ( for a casadi::Function f, with input and output std::vector<std::vector<\double>> x and y). I'm trying to compute them in parallel using std::for_each and std::execution::par. (_Both these options result in segmentation faults for 60 iterations. They work fine however if the number of iterations is reduced to 15_.) :

1. 
-  Call 
f(x,y) -> from this definition -> https://github.com/casadi/casadi/blob/e9072a208fc874622d07a664e98f46e23246ac33/casadi/core/function.cpp#L448

(I found that this option was also using sparsities to assert sizes of inputs and outputs, and it also led to segmentation faults. So, I tried option 2.)

2. 
- Create buffers for input and output of type std::vector<double *> such that,
   buf_in_x[k] = &x[k]
   buf_out_f[k] = &y[k]
- Create work vectors using function sizes :
   iws_f = std::vector<casadi_int>(f.sz_iw());
   ws_f = std::vector <\double> (f.sz_w());
- Call
  f(&buf_in_x.front(), &buf_out_f.front(), &iws_f.front(), &ws_f.front(), 0) -> from this definition -> 
  https://github.com/casadi/casadi/blob/e9072a208fc874622d07a664e98f46e23246ac33/casadi/core/function.cpp#L1454

If possible, could you help fix option **2** or direct me to a better way to make thread-safe function calls? 

Here is part of a stack trace from gdb for option **2** if it helps :
```
#0  0x00007ffff7afe6bc in casadi::casadi_math<double>::fun(unsigned char, double const&, double const*, double*, long long) ()
   from /usr/local/lib/libcasadi.so.3.7   

#1  0x00007ffff7afeaff in casadi::BinaryMX<true, false>::eval(double const**, double**, long long*, double*) const ()
   from /usr/local/lib/libcasadi.so.3.7

#2  0x00007ffff7c3b1d8 in casadi::MXFunction::eval(double const**, double**, long long*, double*, void*) const () from /usr/local/lib/libcasadi.so.3.7

#3  0x00007ffff7bcabfe in casadi::FunctionInternal::eval_gen(double const**, double**, long long*, double*, void*) const ()
   from /usr/local/lib/libcasadi.so.3.7   

#4  0x00007ffff7b88648 in casadi::Function::operator()(double const**, double**, long long*, double*, int) const () from /usr/local/lib/libcasadi.so.3.7

#5  0x00005555556bc4e9 in MPC::<lambda(int)>::operator()(int) const (__closure=0x7fffc53f6120, k=40) at /usr/include/c++/9/bits/stl_iterator.h:803
```

Thanks in advance!

PS, Thanks for a great library!

**Comment by jgillis** [comment]
The thread-safe API is https://github.com/casadi/casadi/blob/e9072a208fc874622d07a664e98f46e23246ac33/casadi/core/function.hpp#L602
Peeking at this internal code may be helpful: https://github.com/casadi/casadi/blob/e9072a208fc874622d07a664e98f46e23246ac33/casadi/core/map.cpp#L424-L458
The last argument, mem, must come from checkout release cfr codegen api: Usage is similar to codegen: https://github.com/casadi/casadi/blob/e9072a208fc874622d07a664e98f46e23246ac33/docs/examples/cplusplus/codegen_usage.cpp#L176

**Reply by jgillis** [reply]
Looks like you were going this way with option 2 already. Make sure each thread runs the call with different res, iw, w and mem arguments. Also, `buf_in_x` must be of size `f.sz_arg()` which is greater than or equal to `f.n_in()`, similar for out.

**Reply by sashankmodali** [reply]
Thanks a lot for your response! I hadn't quite understood how mem was used, so I was using mem=0 for all threads. I will fix this and get back to you. Thanks!
